### PR TITLE
Codeigniter\Entity - Added $deep parameter in toArray and toRawArray

### PR DESCRIPTION
**Pull request by najdanovicivan**
**Description**

Added boolean $deep parameter in toArray and toRawArray methods. This parameters allows casting inner custom and entity objects to array as well

Additionally added :
  - Unit Tests
  - Return Type Declaration for Entity Getters to avoid IDE errors
  - Some Code Style Fixes

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Entity.php [hunk 1]
```diff
@@ -140,11 +140,12 @@ public function fill(array $data = null)
 	 *
 	 * @param boolean $onlyChanged If true, only return values that have changed since object creation
 	 * @param boolean $cast        If true, properties will be casted.
+	 * @param boolean $deep		   If true, inner entities will be casted as array as well.
```
Comment (michalsn) [inline system/Entity.php:143]: Just fixing alignment.
```suggestion
	 * @param boolean $deep        If true, inner entities will be casted as array as well.
```

File: system/Entity.php [hunk 2]
```diff
@@ -163,7 +164,21 @@ public function toArray(bool $onlyChanged = false, bool $cast = true): array
 				continue;
 			}
 
-			$return[$key] = $this->__get($key);
+			$value = $this->__get($key);
```
Comment (michalsn) [inline system/Entity.php:167]: Is it necessary to use an additional variable here? IMO `$return[$key]` would work just fine.

File: system/Entity.php [hunk 3]
```diff
@@ -187,17 +216,35 @@ public function toArray(bool $onlyChanged = false, bool $cast = true): array
...
+	public function toRawArray(bool $onlyChanged = false, bool $deep = false): array
 	{
 		$return = [];
 
 		if (! $onlyChanged)
 		{
-			return $this->attributes;
+			if ($deep)
+			{
+				return array_map (function ($value) use ($onlyChanged, $deep) {
```
Comment (michalsn) [inline system/Entity.php:232]:
```suggestion
				return array_map(function ($value) use ($onlyChanged, $deep) {
```

File: system/Entity.php [hunk 4]
```diff
@@ -187,17 +216,35 @@ public function toArray(bool $onlyChanged = false, bool $cast = true): array
...
+					if ($value instanceof Entity)
+					{
+						$value = $value->toRawArray($onlyChanged, $deep);
+					}
+					else if (is_callable([$value, 'toRawArray']))
```
Comment (michalsn) [inline system/Entity.php:237]:
```suggestion
					elseif (is_callable([$value, 'toRawArray']))
```

File: system/Entity.php [hunk 5]
```diff
@@ -187,17 +216,35 @@ public function toArray(bool $onlyChanged = false, bool $cast = true): array
...
+					{
+						$value = $value->toRawArray();
+					}
+					return $value;
+				}, $this->attributes);
+			}
+			else
+			{
+				return $this->attributes;
+			}
```
Comment (michalsn) [inline system/Entity.php:247]:
```suggestion

			return $this->attributes;
```

File: system/Entity.php [hunk 6]
```diff
@@ -173,7 +188,21 @@ public function toArray(bool $onlyChanged = false, bool $cast = true): array
...
-					$return[$from] = $this->__get($to);
+					$value = $this->__get($to);
+
+					if ($deep)
+					{
+						if ($value instanceof Entity)
+						{
+							$value = $value->toArray($onlyChanged, $cast, $deep);
+						}
+						else if (is_callable([$value, 'toArray']))
```
Comment (michalsn) [inline system/Entity.php:199]:
```suggestion
						elseif (is_callable([$value, 'toArray']))
```

File: system/Entity.php [hunk 7]
```diff
@@ -163,7 +164,21 @@ public function toArray(bool $onlyChanged = false, bool $cast = true): array
...
-			$return[$key] = $this->__get($key);
+			$value = $this->__get($key);
+
+			if ($deep)
+			{
+				if ($value instanceof Entity)
+				{
+					$value = $value->toArray($onlyChanged, $cast, $deep);
+				}
+				else if (is_callable([$value, 'toArray']))
```
Comment (michalsn) [inline system/Entity.php:175]:
```suggestion
				elseif (is_callable([$value, 'toArray']))
```

File: system/Entity.php [hunk 8]
```diff
@@ -173,7 +188,21 @@ public function toArray(bool $onlyChanged = false, bool $cast = true): array
 			{
 				if (array_key_exists($to, $return))
 				{
-					$return[$from] = $this->__get($to);
+					$value = $this->__get($to);
```
Comment (michalsn) [inline system/Entity.php:191]: IMO using `$return[$key]` would be better. No need to introduce an additional variable.

File: system/Entity.php [hunk 9]
```diff
@@ -207,7 +254,19 @@ public function toRawArray(bool $onlyChanged = false): array
...
 			}
 
-			$return[$key] = $this->attributes[$key];
+			if ($deep)
+			{
+				if ($value instanceof Entity)
+				{
+					$value = $value->toRawArray($onlyChanged, $deep);
+				}
+				else if (is_callable([$value, 'toRawArray']))
```
Comment (michalsn) [inline system/Entity.php:263]:
```suggestion
				elseif (is_callable([$value, 'toRawArray']))
```

**Comment by paulbalandan** [comment]
How about renaming the `$deep` parameter to `$recursive`? IMO it is more descriptive of what it does.

**Comment by najdanovicivan** [comment]
> How about renaming the `$deep` parameter to `$recursive`? IMO it is more descriptive of what it does.

@paulbalandan It makes sense to me using either of those names let's see what other think about it.

**Comment by MGatner** [comment]
`$recursive` would be more consistent with other variable names in the framework already.

File: system/Entity.php [hunk 10]
```diff
@@ -207,7 +248,19 @@ public function toRawArray(bool $onlyChanged = false): array
...
```
Comment (michalsn) [inline system/Entity.php:257]: You missed one spot :)
```suggestion
				elseif (is_callable([$value, 'toRawArray']))
```

**Comment by najdanovicivan** [comment]
@michalsn regarding the `else if` and `elseif` can we see to get this rule added in CodeSniffer standard ?

**Pull request merged**